### PR TITLE
A/B Testing - elastic 

### PR DESCRIPTION
**Pull request by nablaone**
This PR add:
- add support for comparison of  Elastic  vs Quesma responses
- add minor improvements related to storing the results 
- move alternative plan handling to a separate file 

Comparison with an Elastic or other external Quesma can be enabled by:
```
  some-index:
    enabled: true
    optimizers:
      elastic_ab_testing:
        enabled: true
        properties:
          mode: "alternative" 
```
Configuration will change in the future.

### CODE REVIEW

File: quesma/ab_testing/collector/processors.go [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+// deAsyncResponse is a processor that processes that removes async "wrapper" from the response
+type deAsyncResponse struct {
+}
+
+func (t *deAsyncResponse) process(in EnrichedResults) (out EnrichedResults, drop bool, err error) {
+
+	deAsync := func(elasticResponse string) (string, error) {
+
+		asJson, err := types.ParseJSON(elasticResponse)
```
Comment (pdelewski) [inline quesma/ab_testing/collector/processors.go:18]: Should not we check `err` as well ?

File: quesma/ab_testing/collector/processors.go [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+		if asJson == nil {
+			return "", err
+		}
+
+		if asJson["response"] == nil {
```
Comment (pdelewski) [inline quesma/ab_testing/collector/processors.go:24]: Maybe 
```suggestion
		if res, ok := asJson["response"]; !ok {
```

File: quesma/quesma/search_alternative.go [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+				alternative = r
+			}
+		}
+
+		bytes, err := body.Bytes()
+		if err != nil {
+			bytes = []byte("error converting body to bytes")
+		}
+
+		toError := func(err error) string {
```
Comment (pdelewski) [inline quesma/quesma/search_alternative.go:81]: `errorToString` would be better name IMO

File: quesma/quesma/search_alternative.go [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+
+	url := q.cfg.Elasticsearch.Url.JoinPath(plan.IndexPattern, "_search").String()
+
+	return alternativePlan, func(ctx context.Context) ([]byte, error) {
+
+		// TODO: add user/pass
+
+		if resp, err := http.Post(url, "application/json", bytes.NewBuffer(requestBody)); err != nil {
+			return nil, err
+		} else {
```
Comment (pdelewski) [inline quesma/quesma/search_alternative.go:157]: This else seems not be needed

File: quesma/quesma/search_alternative.go [hunk 5]
```diff
@@ -0,0 +1,263 @@
...
+
+	q.runExecutePlanAsync(ctx, plan, queryTranslator, table, doneCh, nil)
+
+	response := <-doneCh
+
+	if response.err == nil {
+		if isAsync {
+			asyncResponse := queryparser.SearchToAsyncSearchResponse(response.response, "__quesma_alternative_plan", false, 200)
+			responseBody, err = asyncResponse.Marshal()
+			if err != nil {
```
Comment (pdelewski) [inline quesma/quesma/search_alternative.go:241]: Seems that these checks can be extracted to common place, line 249

File: quesma/quesma/search_alternative.go [hunk 6]
```diff
@@ -0,0 +1,263 @@
...
+	for _, query := range plan.Queries {
+		switch query.Type.AggregationType() {
+		case model.MetricsAggregation, model.BucketAggregation, model.PipelineAggregation:
+			hasAggQuery = true
+		default:
+			queriesWithoutAggr = append(queriesWithoutAggr, query)
+		}
+	}
+
+	if hasAggQuery {
```
[the remark this answers is not in the document]
Comment (nablaone) [reply quesma/quesma/search_alternative.go]: Done

File: quesma/quesma/search_alternative.go (hunk 5 above)
Comment (nablaone) [reply quesma/quesma/search_alternative.go:243]: Do you mean these checks can be shared between the 'main' and the 'alternative' plan execution?

**Review by jakozaur — APPROVED** [review]
Looks good, got minor comment.

File: quesma/ab_testing/collector/fanout.go [hunk 7]
```diff
@@ -14,7 +15,7 @@ type ppPrintFanout struct {
 }
 
 func (t *ppPrintFanout) process(in EnrichedResults) (out EnrichedResults, drop bool, err error) {
-	//pp.Println("A/B Testing FANOUT", in)
+	pp.Println("A/B Testing FANOUT", in)
```
[the remark this answers is not in the document]
Comment (nablaone) [reply quesma/ab_testing/collector/fanout.go]: Yes. It's no longer used.

File: quesma/ab_testing/collector/processors.go [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+	"quesma/quesma/types"
+)
+
+// deAsyncResponse is a processor that processes that removes async "wrapper" from the response
+type deAsyncResponse struct {
+}
+
+func (t *deAsyncResponse) process(in EnrichedResults) (out EnrichedResults, drop bool, err error) {
+
+	deAsync := func(elasticResponse string) (string, error) {
```
[the remark this answers is not in the document]
Comment (nablaone) [reply quesma/ab_testing/collector/processors.go:16]: That's better.

File: quesma/quesma/search_alternative.go (hunk 5 above)
Comment (pdelewski) [reply quesma/quesma/search_alternative.go:243]: I meant that these 3 lines of code can be move below IMO

**Review by pdelewski — APPROVED** [review]

**Pull request merged**